components/api/src/error.rs

```diff
@@ -14,49 +14,116 @@
 
 //! The set of errors for the Exonum API module.
 
+pub use actix_web::http::{
+    header::{self, HeaderName},
+    HeaderMap, StatusCode as HttpStatusCode,
+};
 use failure::Fail;
-use serde::Serialize;
-use std::io;
-
-/// List of possible API errors.
-#[derive(Fail, Debug)]
-pub enum Error {
-    /// Storage error. This type includes errors related to the database, caused
-    /// by, for example, serialization issues.
-    #[fail(display = "Storage error: {}", _0)]
-    Storage(#[cause] failure::Error),
-
-    /// Input/output error. This type includes errors related to files that are not
-    /// a part of the Exonum storage.
-    #[fail(display = "IO error: {}", _0)]
-    Io(#[cause] io::Error),
-
-    /// Bad request. This error occurs when the request contains invalid syntax.
-    #[fail(display = "Bad request: {}", _0)]
-    BadRequest(String),
-
-    /// Moved permanently. This error means that resource existed at the specified
-    /// location, but now is moved to the other place.
-    #[fail(display = "Moved permanently; Location: {}", _0)]
-    MovedPermanently(String),
-
-    /// Gone. This error means that resource existed in the past, but now is not present.
-    #[fail(display = "Gone")]
-    Gone,
-
-    /// Not found. This error occurs when the server cannot locate the requested
-    /// resource.
-    #[fail(display = "Not found: {}", _0)]
-    NotFound(String),
-
-    /// Internal server error. This type can return any internal server error to the user.
-    #[fail(display = "Internal server error: {}", _0)]
-    InternalError(failure::Error),
-
-    /// Unauthorized error. This error occurs when the request lacks valid
-    /// authentication credentials.
-    #[fail(display = "Unauthorized")]
-    Unauthorized,
+use serde::{Deserialize, Serialize};
+
+use std::fmt;
+
+/// API HTTP error struct.
+#[derive(Fail, Debug, PartialEq)]
```

Comment:
I'd be wary of using `PartialEq` in tests: since error headers are not restored, this may lead to unexpected assertion fails. Also, from the point of view of future compatibility having `PartialEq` forces us to implement it for any possible extensions of the type; IMO, it's not worth it. Correspondingly, I'd suggest to compare the necessary fields separately in tests.